### PR TITLE
support for internal git repositories

### PR DESCRIPTION
**Pull request by uturuncoglu**
The manage external is mainly used by the UFSCOMP to checkout sub components and we need to checkout FV3GS code from NOAA's internal git repository called Vlab and gerrit to support standalone FV3GFS application. The following is the actuall git command to checkout the code

git clone --quiet ssh://ufuk.turuncoglu@vlab.ncep.noaa.gov:29418/FV3 FV3GFS

but manage_externals confuses when the url contains user name and port. This PR fixes the issue and allows to checkout code from internal git repository.

**Changes:**
* port and user_name are added as a entry to the Externals.cfg
* user_name can be any string and also special word as GIT_AUTHOR_NAME.
* the GIT_AUTHOR_NAME is replaced by the environment variable with same name

User interface changes?: Yes
Externals.cfg now contains two additional field like port and user_name. It is backward compatible and if they are not set empty string will be used to generate url.

Fixes: No

Testing:
  test removed: N/A
  unit tests: N/A
  system tests: N/A
  manual testing: N/A

### CODE REVIEW

**Comment by uturuncoglu** [comment]
I did not pay attention to the test scripts. I'll try to fix them also because they are complaining about the new fields.

File: manic/repository_git.py
Comment (billsacks) [inline manic/repository_git.py:110]: It would help if you added comments giving examples of what this is doing - both the addition of port and user name. Something along the lines of, "If self._url is XXX and self._port is YYY, then the following will result in url being ZZZ".

File: manic/repository_git.py
Comment (billsacks) [inline manic/repository_git.py:122]: It feels wrong to me to update self._url here: It feels like self._url should be fixed after initialization, otherwise you have the confusing situation that self._url changes over time, depending on whether `self._clone_repo` happens to have been called yet.

I'd suggest introducing a new function on the repository class that builds this full url, including port and user name. Then you can either call it from the `__init__` method, setting `self._url` there, or else you can call it from wherever `self._url` is currently accessed (replacing direct references to `self._url` with calls to this new function, like `self._get_resolved_url()`.

This begs a couple of questions:

(1) Should this fully resolved url be used everywhere where `self._url` currently appears in repository_git.py, or are there some uses that should continue to use the original, unresolved url?

(2) What about repository_svn? Should this functionality be implemented there as well? If not, then please add some error checking that notes that user_name and port cannot be specified for svn repositories.

File: manic/repository_git.py
Comment (billsacks) [inline manic/repository_git.py:114]: Is this always the right logic for where in the url the port should be added, or would it potentially need to be generalized? I really don't know the answer to this myself, but if you think it might need to be generalized in the future, then please add a comment about that.

File: manic/repository_git.py
Comment (billsacks) [inline manic/repository_git.py:271]: Is this logic really correct? It looks to me like, if self._port is not the empty string, then this will treat any url as a match. This seems wrong if there are multiple remotes: won't this just pick the first one all the time?

It's possible that the need for this change would go away with my above suggestions (the comment beginning, "It feels wrong to me to update self._url here").

File: manic/repository.py
Comment (billsacks) [inline manic/repository.py:42]: I believe `XXX is not` is preferred to `not XXX is` (see https://www.python.org/dev/peps/pep-0008/#programming-recommendations)

File: manic/repository_git.py
Comment (billsacks) [reply manic/repository_git.py:271]: Also: I believe `XXX is not` is preferred to `not XXX is` (see https://www.python.org/dev/peps/pep-0008/#programming-recommendations)

**Review by billsacks — CHANGES_REQUESTED** [review]
Thanks for going ahead and implementing this feature that you needed. I could imagine this being useful for others in the future.

In addition to my line comments, I have two overall comments:

(1) Can you please add documentation of this new feature in the help output (which I believe is also copied to README.md by running 'make readme' from the test directory)? In addition to documenting the new optional fields that can be put in Externals.cfg, and giving an example of the resolved URL when using these features (as you do nicely in the PR overview), please also be sure to document the GIT_AUTHOR_NAME environment variable.

(2) Can you please add at least one test (unit or system) covering this new feature?

**Comment by billsacks** [comment]
@gold2718 I have added you as a reviewer, though I understand if you're too swamped for this. I feel like I have reviewed the details of the code reasonably well... although that could always use another set of eyes, I'm mostly hoping for your review of the overall concept and user interface changes.

**Comment by billsacks** [comment]
@uturuncoglu is this still something you want to bring in? Just wondering what your plans are here. If the answer is that you still want this but won't have time to address the review for some amount of time, that's fine, but I just wanted to make sure this doesn't stay open and partially-done indefinitely.

**Comment by uturuncoglu** [comment]
Actually, NOAA is moving their Git repositories from Gerrit (their internal repository) to GitHub and most of the repositories are already moved. So, we no longer need to have this capability and it is priority is low in our side but if you still want to bring in this feature to manage_externals you could keep open this PR and we could make it later but I have no time to add requested changes now.

**Comment by billsacks** [comment]
Okay, given that status, and given that some additional changes might still be needed before this could be merged, I'm going to close this for now. We can always revisit it later if you or others want this functionality.

**Pull request closed without merging**
